### DISCUSSION TITLE
Is there any example on programmatic routes definition?

### DISCUSSION THREAD

**Discussion by turistua** [remix-run/remix#6104] [Q&A] [answered]
Hi everyone,

Files-based routing doesn't work for the project, as there is no prefix for certain URLs (like /products/$slug or /categories/$slug) - each route can be a product or category or anything. Creating a catchAll route is not desired, as it'll require a lot of unnecessary operations before I start renderring data.

I gatsbyjs I can create pages programmatically, like here is docs: https://www.gatsbyjs.com/docs/programmatically-create-pages-from-data/

I assume the same approach can be applied for remix, if I define all these routes programmatically in defineRoutes function. If this is correct, maybe anyone can share the snippet how to do it properly?

Thanks a lot for help!

**Comment by sergiodxa** [comment, marked as answer]
Yes, using the `routes` function in remix.config.js is the way to do it in Remix.

**Reply by turistua** [reply]
@sergiodxa thank you for confirmation! Maybe you could share an example of code how I can fetch the data from the API and create the pages?

**Reply by kiliman** [reply]
The `routes` function is called during the build process. It's async so you can make a fetch call from your API. You essentially return an array of route definitions, which include the URL path, parent, and the route path. If your content is all in `md/mdx`, you can just return those filenames. Otherwise, you must create a valid `jsx/tsx` file with the content. Remix will then compile these files.

Without seeing an example of what your API returns, it's kind of difficult to provide more guidance.

**Reply by turistua** [reply]
@sergiodxa right now when I generate a route, I get an error if want to map a few routes on the same component, like 

`      landingsData.forEach(l => 
        route(l.slug.current, 'templates/landingPage.tsx')
      );
`

saying that the route ID is not unique.

It's not fully documented, so when I dig into the remix sources I found out that the 3rd param for route() can be either a callback or options object. So the fix is following:

`      landingsData.forEach(l => 
        route(l.slug.current, 'templates/landingPage.tsx', {id: l.slug.current})
      );
`

In the same time, with this fix I can't use nested routes anymore. Is it something that can change in the future? I'd prefer to be able to pass both callback and the options object for additional flexibility.

Another possible request for improvement could be to be able to push some context to the route on this stage (like additional param where I can add some object that is accessible from the template - maybe in loader function.

WDYT?

**Reply by kiliman** [reply]
The callback has to be the last argument. The options part is optional.

```ts
route(path, file, () => route(...))
route(path, file, { id: 'some-id' }, () => route(...))
```

https://github.com/remix-run/remix/blob/73e7517ccfc7c524878dd0d3762bd21899630a91/packages/remix-dev/config/routes.ts#L83-L106

As for context, unfortunately, it's not possible with the current implementation. Remember, this is done at build time, and the route manifest is serialized as part of the initial SSR payload (for use by React Router), and RR doesn't support it.

Typically if you need something that is based on the route path, you would fetch that in your loader (database, API, et) and return it for use by the UI component.
